requests_toolbelt/utils/dump.py

```diff
@@ -3,19 +3,116 @@
 
 from requests import compat
 
-
-__all__ = ('dump_response', 'dump_all')
+__all__ = ('dump_response', 'dump_all', 'Sanitizer', 'HeaderSanitizer')
 
 HTTP_VERSIONS = {
     9: b'0.9',
     10: b'1.0',
     11: b'1.1',
 }
 
+#: List of sensitive headers copied from
+#: https://github.com/google/har-sanitizer
+SENSITIVE_HEADERS = {
+    "state",
+    "shdf",
+    "usg",
+    "password",
+    "email",
+    "code",
+    "code_verifier",
+    "client_secret",
+    "client_id",
+    "token",
+    "access_token",
+    "authenticity_token",
+    "id_token",
+    "appid",
+    "challenge",
+    "facetid",
+    "assertion",
+    "fcparams",
+    "serverdata",
+    "authorization",
+    "auth",
+    "x-client-data",
+    "samlrequest",
+    "samlresponse"
+}
+
 _PrefixSettings = collections.namedtuple('PrefixSettings',
                                          ['request', 'response'])
 
 
+class Sanitizer(object):
```

Comment:
There should be a sanitizer interface of sorts that people subclass and override methods without the methods defaulting to unsafe behaviours. Let's have this be a "base class" that raises `NotImplementedError()` exceptions and then an `NoopSanitizer` class that implements this behaviour.

Reply:
I'm not sure I understand how this is unsafe behavior?

Whenever someone needs to implement a sanitizer, they would need to subclass the `NoopSanitizer` instead of the "interface", because it's going to be extremely rare that you want to override all four methods. I'm not sure I see the benefit of splitting the default behavior out from the interface class, but I can do it if you think it is important.

Reply:
I'm confused. Above you talk about not wanting a list of headers to sanitize by default, yet here you are talking about making the default do sanitizing because that is safer. If you don't have a list of headers to sanitize, how do you intend to do "safe-by-default" sanitizing?

If we say that the default is to not do anything (as today, aka unsafe-by-default), then a `HeaderSanitizer` that requires the user to provide a list of headers to sanitize makes sense. The user would the create an instance with a list of headers that they consider sensitive, and pass that as the `sanitizer` argument.

Reply:
Hi there, let me be clear:

If the base sanitizer never sanitizes and silently is a NoOp then it's a bad default. The base sanitizer that users should sub-class should provide errors when trying to use the other methods because the user hasn't defined what they want to do. Otherwise `Sanitizer` in this case is unsafe-by-default. Safe-by-default means making the user think about it in this case.

Further, I'm starting to think this design is too limiting. How can someone say, compose the HeaderSanitizer here and a BodySanitizer elsewhere. They'd have to sub-class and then that becomes a nightmare of copy-pasting code around to make it all play nicely together. Instead, we likely want to accept a list of sanitizers that the user can specify, and call them in the order the user provides.